trace/trace.go

```diff
@@ -107,22 +88,32 @@ func (s traceStep) writeStep(b *bytes.Buffer, formatter string, stepThreshold ti
 type Trace struct {
 	name        string
 	fields      []Field
+	threshold   *time.Duration
 	startTime   time.Time
-	steps       []traceStep
-	nestedTrace []*Trace
+	stepsTraces []stepTrace
 }
 
 func (t *Trace) time() time.Time {
 	return t.startTime
 }
 
-func (t *Trace) writeStep(b *bytes.Buffer, formatter string, stepThreshold time.Duration,
+func (t *Trace) writeStep(b *bytes.Buffer, formatter string, parentTrace *Trace, stepThreshold time.Duration,
 	lastStepTime time.Time) time.Time {
-	b.WriteString(fmt.Sprintf("%v[", formatter))
-	writeMainLog(b, t.name, t.TotalTime(), t.startTime, t.fields)
-	_ = writeTrace(b, t, formatter+" ", stepThreshold)
-	b.WriteString("]")
-	return time.Time{}
+	stepDuration := t.startTime.Sub(lastStepTime)
+	if stepThreshold == 0 || stepDuration > stepThreshold || klog.V(4) {
+		if t.threshold != nil {
+			stepThreshold = *parentTrace.threshold - *t.threshold
+			limitThreshold := *parentTrace.threshold / 4
+			if stepThreshold < limitThreshold {
+				stepThreshold = limitThreshold
+			}
```

Comment:
I'm not following the logic behind this math (and it look different than what I had expected based on discussion). Explain somewhere?

Reply:
>Subtract the nested trace thresholds from the threshold of the current trace to calculate a >remaining threshold for the steps of the current trace
line 105: subtracts the threshold of the trace from that of the parent trace

>less than 1/4 of the current trace threshold is used to calculate the stepThreshold for the >steps that are direct children of the trace
line 106 - 108: if the resulting threshold(from the calculation) is less than 1/4 of theparent trace. It uses the 1/4 of the parent trace instead

Did I miss it somewhere?

Reply:
Just to capture what we discussed offline:

Say we have something like:

```
Trace1: threshold=1000ms (provided by call to LogIfLong)
  step1.1
  Trace2: threshold=600ms (provided by call to LogIfLong)
  step1.2
  Trace3: (no threshold provided)
```

My thought was that the effective thresholds of step1.1, step1.2 and Trace3 might be calculated based on the threshold of Trace1 and Trace2. Since Trace2 accounts for 600ms of the 1000ms of Trace1, I would expect the remaining time to then be divided up evenly, e.g.:

```
Trace1: threshold=1000ms
  step1.1: stepThreshold=100ms
  Trace2: threshold=600ms
  step1.2: stepThreshold=100ms
  Trace3: threshold=100ms
```

Where there is 100ms still left for Trace1 directly (which is why, I assume we do the `+1` in `len(steps)+1` already).